go/chat/attachment_crypt.go

```diff
@@ -1,45 +1,112 @@
 package chat
 
-import "io"
+import (
+	"crypto/rand"
+	"errors"
+	"io"
+
+	"github.com/agl/ed25519"
+	"github.com/keybase/client/go/chat/signencrypt"
+	"github.com/keybase/client/go/libkb"
+)
 
 type Encrypter interface {
 	// EncryptedLen returns the number of bytes that the ciphertext of
 	// size plaintext bytes will be.
 	EncryptedLen(size int) int
 
 	// Encrypt takes a plaintext reader and returns a ciphertext reader.
-	Encrypt(plaintext io.Reader) (ciphertext io.Reader)
+	// It generates new keys every time it is called.
+	Encrypt(plaintext io.Reader) (ciphertext io.Reader, err error)
+
+	// EncryptKey returns the ephemeral key that was used during Encrypt.
+	EncryptKey() []byte
 
-	// Key returns the ephemeral key that was used during Encrypt.
-	Key() []byte
+	// VerifyKey returns the public portion of the signing key that
+	// can be used for signature verification.
+	VerifyKey() []byte
 }
 
 type Decrypter interface {
-	// Decrypt takes a ciphertext reader and a key and returns a
-	// plaintext reader.
-	Decrypt(ciphertext io.Reader, key []byte) (plaintext io.Reader, err error)
+	// Decrypt takes a ciphertext reader, encryption and verify keys.
+	// It returns a plaintext reader.
+	Decrypt(ciphertext io.Reader, encKey, verifyKey []byte) (plaintext io.Reader)
 }
 
-// PassThrough implements Encrypter and Decrypter.  It does no encryption
-// or decryption.
-type PassThrough struct{}
+var nonce signencrypt.Nonce
 
-func NewPassThrough() *PassThrough {
-	return &PassThrough{}
+func init() {
+	var n [signencrypt.NonceSize]byte
+	copy(n[:], "kbchatattachment")
+	nonce = &n
 }
 
-func (p *PassThrough) EncryptedLen(size int) int {
-	return size
+type SignEncrypter struct {
+	encKey    signencrypt.SecretboxKey
+	signKey   signencrypt.SignKey
+	verifyKey signencrypt.VerifyKey
 }
 
-func (p *PassThrough) Encrypt(r io.Reader) io.Reader {
-	return r
+func NewSignEncrypter() *SignEncrypter {
+	return &SignEncrypter{}
 }
 
-func (p *PassThrough) Key() []byte {
+func (s *SignEncrypter) EncryptedLen(size int) int {
+	return signencrypt.GetSealedSize(size)
+}
+
+func (s *SignEncrypter) Encrypt(r io.Reader) (io.Reader, error) {
+	if err := s.makeKeys(); err != nil {
+		return nil, err
+	}
+	return signencrypt.NewEncodingReader(s.encKey, s.signKey, nonce, r), nil
+}
+
+func (s *SignEncrypter) EncryptKey() []byte {
+	return []byte((*s.encKey)[:])
+}
+
+func (s *SignEncrypter) VerifyKey() []byte {
+	return []byte((*s.verifyKey)[:])
+}
+
+func (s *SignEncrypter) makeKeys() error {
+	var encKey [signencrypt.SecretboxKeySize]byte
+	n, err := rand.Read(encKey[:])
+	if err != nil {
+		return err
+	}
+	if n != signencrypt.SecretboxKeySize {
+		return errors.New("failed to rand.Read the correct number of bytes")
+	}
```

Comment:
This is another one of those checks that should be impossible I think? But again, doesn't hurt to leave it.